plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/refactoring/rename/wizard/RenamePresenter.java

```diff
@@ -225,214 +219,131 @@ private void setEditorFocus() {
+            arg -> {
+              notificationManager.notify(
+                  locale.failedToRename(), arg.getMessage(), FAIL, FLOAT_MODE);
             });
   }
 
-  private Promise<ChangeCreationResult> prepareRenameChanges(final RefactoringSession session) {
-    RenameSettings renameSettings = createRenameSettingsDto(session);
+  private RenameParams createRenameParams(RenameSettings renameSettings) {
+    RenameParams renameParams = dtoFactory.createDto(RenameParams.class);
+    renameParams.setNewName(view.getNewName());
+
+    if (RefactoredItemType.JAVA_ELEMENT.equals(refactorInfo.getRefactoredItemType())) {
+      EditorPartPresenter activeEditor = editorAgent.getActiveEditor();
+      VirtualFile file = activeEditor.getEditorInput().getFile();
+      TextDocumentIdentifier textDocumentIdentifier = dtoBuildHelper.createTDI(file);
+      renameParams.setTextDocument(textDocumentIdentifier);
+
+      TextPosition cursorPosition = ((TextEditor) activeEditor).getCursorPosition();
+      org.eclipse.lsp4j.Position position = dtoFactory.createDto(org.eclipse.lsp4j.Position.class);
+      position.setCharacter(cursorPosition.getCharacter());
+      position.setLine(cursorPosition.getLine());
+      renameParams.setPosition(position);
+      renameSettings.setRenameType(RenameType.JAVA_ELEMENT);
+    } else if (RefactoredItemType.COMPILATION_UNIT.equals(refactorInfo.getRefactoredItemType())) {
+      renameSettings.setRenameType(RenameType.COMPILATION_UNIT);
+      TextDocumentIdentifier textDocumentIdentifier =
+          dtoFactory.createDto(TextDocumentIdentifier.class);
+      textDocumentIdentifier.setUri(refactorInfo.getResources()[0].getLocation().toString());
+      renameParams.setTextDocument(textDocumentIdentifier);
+    } else if (RefactoredItemType.PACKAGE.equals(refactorInfo.getRefactoredItemType())) {
+      renameSettings.setRenameType(RenameType.PACKAGE);
+      TextDocumentIdentifier textDocumentIdentifier =
+          dtoFactory.createDto(TextDocumentIdentifier.class);
+      textDocumentIdentifier.setUri(refactorInfo.getResources()[0].getLocation().toString());
+      renameParams.setTextDocument(textDocumentIdentifier);
+    }
+    return renameParams;
+  }
 
-    return refactorService
-        .setRenameSettings(renameSettings)
-        .thenPromise(
-            new Function<Void, Promise<ChangeCreationResult>>() {
-              @Override
-              public Promise<ChangeCreationResult> apply(Void arg) throws FunctionException {
-                return refactorService.createChange(session);
-              }
-            });
+  private void applyRefactoring(CheWorkspaceEdit workspaceEdit) {
+    view.close();
+    WorkspaceEdit edit = new WorkspaceEdit();
+    edit.setChanges(workspaceEdit.getChanges());
+    edit.setResourceChanges(new LinkedList<>());
+    for (ResourceChange resourceChange : workspaceEdit.getResourceChanges()) {
+      edit.getResourceChanges().add(Either.forLeft(resourceChange));
```

Comment:
Out of curiosity...I thought sending WorkspaceChange over the wire did not work? What has changed?

Reply:
Sending `List<Either<ResourceChange, TextEdit>>` which is described in `org.eclipse.lsp4j.WorkspaceEdit` doesn't work. 
So here `workspaceEdit` is an instance of custom object (`org.eclipse.che.jdt.ls.extension.api.dto.CheWorkspaceEdit`) where I already have all actual changes and I take all changes from it and put them to the `edit` (`org.eclipse.lsp4j.WorkspaceEdit`) to sent it to class which applies `org.eclipse.lsp4j.WorkspaceEdit` (`org.eclipse.che.plugin.languageserver.ide.editor.quickassist.ApplyWorkspaceEditAction#ApplyWorkspaceEditAction`)